### PR TITLE
Feature/gwdc 74 different warehouses

### PR DESCRIPTION
**Pull request by Martin-Welte**
# TL;DR

- adds action to orders to change to fullfillment state, to make it easier for shop owners to trigger fullfilllment
- adds new options tab
- adds new shipping method, where warehouse id can be set
- updates tests

### CODE REVIEW

File: includes/Actions/SendOrderAction.php [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        $this->settingsRepository = $settingsRepository;
+    }
+
+    /**
+     * Sets the OrderStatus to the fulfillment state.
+     * This will trigger the order to be sent to GBW.
+     *
+     * @param object $order The order (WC_Order) to be sent to GBW.
+     */
+    public function sendOrderToGbw(object $order)
```
Comment (philipp-hemmer) [inline includes/Actions/SendOrderAction.php:52]: i would place this function after addActions()

File: includes/ShippingMethods/GBWShippingMethod.php [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    /**
+     * Calculate the shipping rate.
+     *
+     * @param array $package The package to calculate the shipping for.
+     */
+    public function calculate_shipping($package = array()): void
+    {
+        $rate = array(
+            'label' => $this->title,
+            'cost' => '0',
```
Comment (philipp-hemmer) [inline includes/ShippingMethods/GBWShippingMethod.php:89]: interesting that this is a string

File: templates/options/boolean.twig [hunk 3]
```diff
@@ -0,0 +1,4 @@
+<input type="checkbox" name="{{ option.slug }}" {{ option.value == true ? 'checked' : '' }}><br/>
```
Comment (philipp-hemmer) [inline templates/options/boolean.twig]: i think `option.value === true` is possible here

File: includes/ShippingMethods/GBWShippingMethod.php (hunk 2 above)
Comment (Martin-Welte) [reply includes/ShippingMethods/GBWShippingMethod.php:89]: yes, agree. but that is wordpress development for you.

File: templates/options/boolean.twig (hunk 3 above)
Comment (Martin-Welte) [reply templates/options/boolean.twig]: no, not in twig. there it is called is same as.

File: tests/Unit/LogisticsOrderFactoryTest.php [hunk 4]
```diff
@@ -253,6 +270,33 @@ public function test_it_adds_a_logistics_product()
...
     }
 
+    /**
+     * @runInSeparateProcess
+     * @preserveGlobalState disabled
+     */
+    public function test_it_adds_the_warehouse_id_if_a_gbw_shipping_method_is_used()
+    {
+        // warehouse id has to be between 10 and 12 characters
+        $warehouseID = rand_str(11);
```
Comment (mattcrn) [inline tests/Unit/LogisticsOrderFactoryTest.php]: I wouldn't use random in Unit tests

**Review by mattcrn — COMMENTED** [review]
Looks good just 1 comment

File: templates/options/boolean.twig (hunk 3 above)
Comment (philipp-hemmer) [reply templates/options/boolean.twig]: can we use `option.value is same as(true)`?

Comment (Martin-Welte) [reply templates/options/boolean.twig]: not easily, i need to typecast when getting the value from the database. But it's a good point.

**Review by mattcrn — APPROVED** [review]

**Pull request merged**